hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,478 @@
+                    -> (Parameter device dtype '[4 * hiddenSize, numDirections * hiddenSize])
+                    -> (Parameter device dtype '[4 * hiddenSize, hiddenSize])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> (Parameter device dtype '[4 * hiddenSize ])
+                    -> LSTMParams device dtype numDirections inputDim hiddenSize (numLayers + 1)
+
+--  TODO: Generics? see https://gist.github.com/RyanGlScott/71d9f933e823b4a03f99de54d4b94d51
+
+-- A specialized singlton helper for initializing parameters
+class (KnownNat n) => LayerSong (n :: Nat) where
+    singLayerSing :: (RandDTypeIsValid device dtype, KnownDevice device, KnownDType dtype, KnownNat numDirections, KnownNat inputDim, KnownNat hiddenSize)
+        =>  IO (LSTMParams device dtype numDirections inputDim hiddenSize n)
+
+instance {-# OVERLAPS #-} LayerSong 1 where
+    singLayerSing =
+        LSTMLayer1
+            <$> (makeIndependent =<< xavierUniormLSTM)
+            <*> (makeIndependent =<< xavierUniormLSTM)
+            <*> (makeIndependent =<< (pure zeros))
+            <*> (makeIndependent =<< (pure zeros))
+
+instance {-# OVERLAPPABLE #-} (KnownNat n, KnownNat m, LayerSong n, m ~ (n + 1), 1 <= n) => LayerSong m where
+    singLayerSing =
+        LSTMLayerK
+            <$> singLayerSing
+            <*> (makeIndependent =<< xavierUniormLSTM)
+            <*> (makeIndependent =<< xavierUniormLSTM)
+            <*> (makeIndependent =<< (pure zeros))
+            <*> (makeIndependent =<< (pure zeros))
+
+instance A.Parameterized (LSTMParams device dtype numDirections inputDim hiddenSize numLayers) where
+    flattenParameters (LSTMLayer1 a b c d) =
+        A.flattenParameters a
+            <> A.flattenParameters b
+            <> A.flattenParameters c
+            <> A.flattenParameters d
+    flattenParameters (LSTMLayerK l a b c d) =
+        A.flattenParameters a
+            <> A.flattenParameters b
+            <> A.flattenParameters c
+            <> A.flattenParameters d
+            <> A.flattenParameters l
+    replaceOwnParameters (LSTMLayer1 a b c d) =
+        LSTMLayer1
+            <$> A.replaceOwnParameters a
+            <*> A.replaceOwnParameters b
+            <*> A.replaceOwnParameters c
+            <*> A.replaceOwnParameters d
+    replaceOwnParameters (LSTMLayerK l a b c d) =
+        LSTMLayerK
+            <$> A.replaceOwnParameters l
+            <*> A.replaceOwnParameters a
+            <*> A.replaceOwnParameters b
+            <*> A.replaceOwnParameters c
+            <*> A.replaceOwnParameters d
+
+-- | The directional specification of the LSTM
+--
+data Directionality =
+    Bidirectional -- ^ Forward and backward along the sequential axis using independant parameters for each.
+    | Unidirectional -- ^ Forward along the sequential axis.
+    deriving (Show, Generic) -- TODO:  We could also have BidirectionalTied weights.
+
+type family NumberOfDirections a where
```

Comment:
could you annotate the kind here please, so that a `NumberOfDirections dir` constraint always forces `dir :: Directionality`?

Reply:
same for the return kind, i.e. `Nat`

Reply:
Right, done.